undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```

Comment:
A single boolean flag (`canMerge`) will not suffice, since we can no longer forget that the user called `preventMerge` once the next change (`c2` in my previous comment) arrives. One way to store in the queue `Either<Barrier, C>`, where left case means to prevent merge of changes to the left and right of the "barrier". (`Barrier` could actually be `Void`, or, if we insist that user's changes are not `null`, we could just use misuse `null` to mean the barrier, to avoid too much boxing.